addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java

```diff
@@ -143,49 +175,132 @@ private void changeState(final InternalLedState newState) throws IOException {
+        protected InternalLedState currentFadeState;
+        protected InternalLedState toState;
+        protected InternalLedState fromState;
+
+        // After the last update keep the communication channel open for XXX ms
+        protected long keepCommOpenForMS = 1000;
+        // Keep track of when last communication happened
+        protected long lastComTime = 0;
+        // Function that will be called to update LED
+        protected final Function<InternalLedState, Boolean> ledSender;
+
+        protected final Lock lock = new ReentrantLock();
+
+        public LEDFaderRunner(InternalLedState fromState, InternalLedState toState, int steps, int keepCommOpenForMS, Function<InternalLedState, Boolean> ledSender) {
+            this.currentFadeState = fromState;
+            this.fromState = fromState;
+            this.toState = toState;
+            this.steps = steps;
+            this.ledSender = ledSender;
+            this.keepCommOpenForMS = keepCommOpenForMS;
+        }
 
-                        InternalLedState fadeState = currentTargetState;
+        public void setToState(InternalLedState toState) {
+            lock.lock();
+            this.fromState = currentFadeState;
+            this.toState = toState;
+            this.fadeStepper = 1;
+            lock.unlock();
+        }
 
-                        for (int i = 1; i <= fadeSteps && keepFading; i++) {
-                            long lastTime = System.nanoTime();
-                            fadeState = currentTargetState.fade(newState, (double) i / fadeSteps);
-                            logger.debug("fadeState: {}", fadeState);
+        public void run() {
+            lock.lock();
+            if (fadeStepper <= steps) {
+                currentFadeState = fromState.fade(toState, (double) fadeStepper / steps);
+                lock.unlock();
 
-                            sendLEDData(fadeState, outputStream);
+                logger.debug("currentFadeState: {}", currentFadeState);
+                if (!ledSender.apply(currentFadeState)) {
+                    logger.warn("Failed sending at step {}", fadeStepper);
+                    throw new RuntimeException("Failed sending at step " + fadeStepper);
+                }
+                lastComTime = System.currentTimeMillis();
+            } else {
+                lock.unlock();
+                if (lastComTime < (System.currentTimeMillis() - keepCommOpenForMS)) {
+                    throw new RuntimeException("Reached end step");
+                }
+            }
+            fadeStepper++;
+        }
+    }
 
-                            busySleep(fadeDurationInMs / fadeSteps, lastTime);
+    synchronized private void fadeToState(final InternalLedState newTargetState) throws IOException {
+
+        if (ledUpdateSyncSemaphore.tryAcquire(1)) {
+            final Future<?> future;
+            final Socket socket;
+
+            // Create and Execute a new LED Fader
+            try {
```

Comment:
Do you close this socket anywhere, I think you should use the try-with-resources here:

https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html